lua/nvim-treesitter/node-movement.lua

```diff
@@ -0,0 +1,125 @@
+local api = vim.api
+local parsers = require'nvim-treesitter.parsers'
+local M = {}
+
+
+M.NodeMovementKind = {
+  up = 'up',
+  down = 'down',
+  left = 'left',
+  right = 'right',
+}
+
+M.current_node = {}
+
+local function node_start_to_vim(node)
+  if not node then return end
+
+  local row, col = node:start()
+  local exec_command = string.format('call cursor(%d, %d)', row+1, col+1)
+
+  api.nvim_exec(exec_command, false)
+end
+
+
+local function get_previous_node(node)
+  local destination_node
+  local parent = node:parent()
+  if parent then
+    local found_pos = 0
+    for i = 0,parent:named_child_count()-1,1 do
+      if parent:named_child(i) == node then
+        found_pos = i
+        break
+      end
+    end
+    if 0 <= found_pos - 1 then
+      destination_node = parent:named_child(found_pos - 1)
+    else
+      local previous_node = get_previous_node(node:parent())
+      if previous_node and previous_node:named_child_count() > 0 then
+        destination_node = previous_node:named_child(previous_node:named_child_count() - 1)
+      elseif previous_node then
+        destination_node = previous_node
+      end
+    end
+  end
+  return destination_node
+end
+
+
+local function get_next_node(node)
+  local destination_node
+  local parent = node:parent()
+
+  if parent then
+    local found_pos = 0
+    for i = 0,parent:named_child_count()-1,1 do
+      if parent:named_child(i) == node then
+        found_pos = i
+        break
+      end
+    end
+    if parent:named_child_count() > found_pos + 1 then
+      destination_node = parent:named_child(found_pos + 1)
+    else
+      local next_node = get_next_node(node:parent())
+      if next_node and next_node:named_child_count() > 0 then
+        destination_node = next_node:named_child(0)
+      elseif next_node then
+        destination_node = next_node
+      end
+    end
+  end
+  return destination_node
+end
```

Comment:
Could you move those two functions in the utils file ? It might be reused by other modules

Reply:
Ok. I did that I tried to hide the extra magic (module-specific behavior) into optional parameters.